### Projekt Opgave 1-2 med Breast Cancer Datasæt: Udforskning af data
1. Udforsk datasættene på forskellige måder (indlæs data, udskriv noget data, tjek datas dimensioner / størrelse osv. osv.).
2. Udarbejd en (Python) applikation som kan indlæse og vise datasættene

### Imports

In [17]:
from sklearn.datasets import load_breast_cancer
import pandas as pd
import numpy as np

### Indlæsning af Data for Breast Cancer

In [38]:
breast_cancer = load_breast_cancer()

df = pd.DataFrame(data=breast_cancer['data'], columns=breast_cancer['feature_names'])
df['target'] = breast_cancer['target']
df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  target  
0                  0.2654          0.4601                  0.11890       0  
1                  0.1860          0.2750                  0.08902       0  
2                  0.2430          0.3613                  0.08758       0  
3                  0.2575          0.6638                  0.17300       0  
4                  0.1625          0.2364                  0.076

### Visning af de første 5 elementer i DataFramen

In [39]:
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

### Blande i datasættet (Shuffle)

Man kan shuffle i datasættet. Man kan selv vælge om man vil shuffle i dele af den (med frac) (så fx. halvdelen (0.5) (5%) eller et givet tal som fx 0.78 (78%)) eller hele datasættet.

Man kan også bare give sample et tal og det bliver så antallet der bliver shufflet i.

In [46]:
df.sample(2)

# eller
print("Antal Elementer i datasættet: %s" %(len(df)))
df.sample(frac=0.5)

Antal Elementer i datasættet: 569


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
276       11.330         14.16           71.79      396.6          0.09379   
316       12.180         14.08           77.25      461.4          0.07734   
250       20.940         23.56          138.90     1364.0          0.10070   
66         9.465         21.01           60.11      269.4          0.10440   
223       15.750         20.25          102.60      761.3          0.10250   
..           ...           ...             ...        ...              ...   
191       12.770         21.41           82.02      507.4          0.08749   
275       11.890         17.36           76.20      435.6          0.12250   
471       12.040         28.14           76.85      449.9          0.08752   
107       12.360         18.54           79.01      466.7          0.08477   
420       11.570         19.04           74.20      409.7          0.08546   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
276           0.03872        0.001487             0.003333         0.1954   
316           0.03212        0.011230             0.005051         0.1673   
250           0.16060        0.271200             0.131000         0.2205   
66            0.07773        0.021720             0.015040         0.1717   
223           0.12040        0.114700             0.064620         0.1935   
..                ...             ...                  ...            ...   
191           0.06601        0.031120             0.028640         0.1694   
275           0.07210        0.059290             0.074040         0.2015   
471           0.06000        0.023670             0.023770         0.1854   
107           0.06815        0.026430             0.019210         0.1602   
420           0.07722        0.054850             0.014280         0.2031   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
276                 0.05821  ...          18.99            77.37       458.0   
316                 0.05649  ...          16.47            81.60       513.1   
250                 0.05898  ...          27.00           165.30      2010.0   
66                  0.06899  ...          31.56            67.03       330.7   
223                 0.06303  ...          30.29           125.90      1088.0   
..                      ...  ...            ...              ...         ...   
191                 0.06287  ...          23.50            89.04       579.5   
275                 0.05875  ...          18.99            79.46       472.4   
471                 0.05698  ...          33.33            87.24       567.6   
107                 0.06066  ...          27.49            85.56       544.1   
420                 0.06267  ...          26.98            86.43       520.5   

     worst smoothness  worst compactness  worst concavity  \
276           0.12590            0.07348         0.004955   
316           0.10010            0.05332         0.041160   
250           0.12110            0.31720         0.699100   
66            0.15480            0.16640         0.094120   
223           0.15520            0.44800         0.397600   
..                ...                ...              ...   
191           0.09388            0.08978         0.051860   
275           0.13590            0.08368         0.071530   
471           0.10410            0.09726         0.055240   
107           0.11840            0.19630         0.193700   
420           0.12490            0.19370         0.256000   

     worst concave points  worst symmetry  worst fractal dimension  target  
276               0.01111          0.2758                  0.06386       1  
316               0.01852          0.2293                  0.06037       1  
250               0.21050          0.3126                  0.07849       0  
66                0.06517          0.2878                  0.09211       1  
223               0.14790          0.3993                  0.106

### Middelværdier, Gennemsnit og mere i datasættet

In [61]:
print("Gennemsnittet af 'mean radius' er %s" %(df['mean radius'].mean()))

amount = len(df[df['target'] == 1])
print(f'Antallet af patienter med cancer: {amount}')


Gennemsnittet af 'mean radius' er 14.127291739894552
Antallet af patienter med cancer: 357


### Dimensioner

Kan aflæses sådan: Den yderste dimension har 569 arrays med 31 elementer hver.

In [81]:
df.shape

(569, 31)

In [91]:

print("All features in Breast Cancer dataset")

#Grunden til at jeg skriver -1 er fordi den sidste kolonne er selve Target, som ikke er en feature, men den label man tester med.
for feature in df.columns[:-1]:
    print(feature)

All features in Breast Cancer dataset
mean radius
mean texture
mean perimeter
mean area
mean smoothness
mean compactness
mean concavity
mean concave points
mean symmetry
mean fractal dimension
radius error
texture error
perimeter error
area error
smoothness error
compactness error
concavity error
concave points error
symmetry error
fractal dimension error
worst radius
worst texture
worst perimeter
worst area
worst smoothness
worst compactness
worst concavity
worst concave points
worst symmetry
worst fractal dimension


### Generering af beskrivende statistik (Descriptive Statistics)

Ved at gøre brug af `.describe()` metoden kan man få en beskrivende resume. Som følgende:

In [97]:
df['worst symmetry'].describe()

count    569.000000
mean       0.290076
std        0.061867
min        0.156500
25%        0.250400
50%        0.282200
75%        0.317900
max        0.663800
Name: worst symmetry, dtype: float64

In [98]:
df.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension      target  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]